src/main/scala/medeia/decoder/BsonKeyDecoder.scala

```diff
@@ -0,0 +1,42 @@
+package medeia.decoder
+
+import java.util.UUID
+
+import cats.Functor
+import cats.syntax.either._
+import cats.data.EitherNec
+import medeia.decoder.BsonDecoderError.FieldParseError
+
+trait BsonKeyDecoder[A] { self =>
+  def decode(string: String): EitherNec[BsonDecoderError, A]
+
+  def map[B](f: A => B): BsonKeyDecoder[B] = (string: String) => self.decode(string).map(f)
+}
+
+object BsonKeyDecoder extends DefaultBsonKeyDecoderInstances {
+  def apply[A: BsonKeyDecoder]: BsonKeyDecoder[A] = implicitly
+
+  def decode[A: BsonKeyDecoder](string: String): EitherNec[BsonDecoderError, A] = {
+    BsonKeyDecoder[A].decode(string)
+  }
+
+  implicit val bsonDecoderFunctor: Functor[BsonDecoder] = new Functor[BsonDecoder] {
+    override def map[A, B](fa: BsonDecoder[A])(f: A => B): BsonDecoder[B] = fa.map(f)
+  }
+}
+
+trait DefaultBsonKeyDecoderInstances {
+  implicit val stringDecoder: BsonKeyDecoder[String] = string => Right(string)
+
+  implicit val intDecoder: BsonKeyDecoder[Int] = string =>
```

Comment:
nitpick :D maybe we should rename the argument from `string` to something like `key`

Reply:
done